easybuild/tools/filetools.py

```diff
@@ -1360,35 +1366,37 @@ def apply_regex_substitutions(paths, regex_subs, backup='.orig.eb'):
         for regex, subtxt in regex_subs:
             compiled_regex_subs.append((re.compile(regex), subtxt))
 
-        if backup:
-            backup_ext = backup
-        else:
-            # no (persistent) backup file is created if empty string value is passed to 'backup' in fileinput.input
-            backup_ext = ''
-
         for path in paths:
             try:
                 # make sure that file can be opened in text mode;
                 # it's possible this fails with UnicodeDecodeError when running EasyBuild with Python 3
                 try:
-                    with open(path, 'r') as fp:
-                        _ = fp.read()
+                    with _open(path, 'r') as fp:
+                        txt_utf8 = fp.read()
                 except UnicodeDecodeError as err:
                     _log.info("Encountered UnicodeDecodeError when opening %s in text mode: %s", path, err)
                     path_backup = back_up_file(path)
                     _log.info("Editing %s to strip out non-UTF-8 characters (backup at %s)", path, path_backup)
                     txt = read_file(path, mode='rb')
                     txt_utf8 = txt.decode(encoding='utf-8', errors='replace')
+                    del txt
                     write_file(path, txt_utf8)
 
-                for line_id, line in enumerate(fileinput.input(path, inplace=1, backup=backup_ext)):
-                    for regex, subtxt in compiled_regex_subs:
-                        match = regex.search(line)
-                        if match:
-                            origtxt = match.group(0)
-                            _log.info("Replacing line %d in %s: '%s' -> '%s'", (line_id + 1), path, origtxt, subtxt)
-                        line = regex.sub(subtxt, line)
-                    sys.stdout.write(line)
+                if backup:
+                    copy_file(path, path + backup)
+                with _open(path, 'w') as out_file:
+                    lines = txt_utf8.split('\n')
+                    del txt_utf8
```

Comment:
Why the `del` here already? Just let it go out of scope so garbage collector will pick it up?

This looks weird/dangerous to me, so if you have a good reason for it, you should add a comment.

Reply:
The scope is the function and the file read might be large. We basically copy the contents multiple times: input string, list and joined list. I wanted to keep the scope of that short to limit the amount of memory used similar to what `fileinput` did: Processing the file line-wise not as a whole. We don't do that anymore but can at least avoid having 3 copies in memory or accidentally using the wrong variable, so this is rather safe than dangerous

Note that the `del` is safe: It is simply that: It ends the variables lifetime prior to its scope, may lead to freeing its resources by the GC and if you were to use that variable you'd get an error "variable does not exist".
As `del` is a core Python feature I wouldn't add a comment here. The statement alone says "I'm done with this variable"

Reply:
OK, I'd expect it to be fine too, but it's still a pattern you don't see very often. Let's leave it as is, thanks for clarifying...